src/wallet/walletdb.cpp

```diff
@@ -1248,8 +1248,13 @@ bool CWalletDB::Recover(CDBEnv& dbenv, const std::string& filename, bool fOnlyKe
             CDataStream ssKey(row.first, SER_DISK, CLIENT_VERSION);
             CDataStream ssValue(row.second, SER_DISK, CLIENT_VERSION);
             string strType, strErr;
-            bool fReadOK = ReadKeyValue(&dummyWallet, ssKey, ssValue,
+            bool fReadOK;
+            {
+                // Required in LoadKeyMetadata():
+                LOCK(dummyWallet.cs_wallet);
+                fReadOK = ReadKeyValue(&dummyWallet, ssKey, ssValue,
```

Comment:
If ReadKeyValue needs the lock, is there a reason that it doesn't take that lock internally? That seems like it'd be safer.

Reply:
I'll check, all I'm doing now is cherry-picking commits. One thing I'm trying to remember to do, and reviewers can help with this (and it's pretty easy) is to look at the last part of the upstream PR (in this case, 7229) to see if any later bitcoin/bitcoin PRs refer to this PR (7229). If this PR (7229) was later reverted, the later PR that does that will certainly reference it, and doing that adds that reference (a back pointer, if you will) to the current (7229) PR. (Sorry I have to keep writing 7229, I just wanted to be clear which PR I'm talking about.) If you see no such back-pointer reference, then this PR (7229) is very likely a good change, even if the code was later refactored and latest master looks nothing like this (for example, in this case, maybe the locking was moved to within `ReadKeyValue()`). We still want this change and, I would say, in the form it was cherry-picked.

Reply:
That back-pointer should be visible in the github UI if there is one, they automatically record mentions. I don't see any back-references within bitcoin/bitcoin, though there is this other back-reference from another project https://github.com/bitcoinxt/bitcoinxt/pull/301/files which says it adds tests related to this change? I haven't looked at what those tests are actually doing but maybe they're worth a glance.

Reply:
The changes made by the two test commits have already been made in zcash (but without cherry-picks) or no longer apply. In general, if there are test changes that go with these locking fixes, I'll do my best to make sure to cherry-pick them too.